File: java_solutions/198.house-robber.java
/*
 * @lc app=leetcode id=198 lang=java
 *
 * [198] House Robber
 */

class Solution {
    public int rob(int[] nums) {
        return Arrays.stream(nums)
            .boxed()
            .reduce(new int[]{0, 0}, 
                (state, n) -> new int[]{state[1], Math.max(state[1], state[0] + n)}, 
                (a, b) -> a)[1];
    }
}

File: kotlin_solutions/198.house-robber.kt
/*
 * @lc app=leetcode id=198 lang=kotlin
 *
 * [198] House Robber
 */

class Solution { fun rob(nums: IntArray): Int = nums.fold(0 to 0) { (prev2, prev1), num -> prev1 to maxOf(prev1, prev2 + num) }.second }
